2016/proposals/narayan-aditya-visualizer.md

```diff
@@ -0,0 +1,244 @@
+ * Backend  : [Django Rest Framework]
+ * Scheduler: [Cron]
+ * Database : [PostgreSQL]
+ * Frontend : [Charts.js] and [Bootstrap]
+
+### Backend - Django Application
+
+The survey visualizer application would be made using
+[Django Rest Framework] (henceforth called DRF) over [Django]
+to allow easier integration with [AMY] when required.
+
+The tentative API structure exposed by DRF is:
+ * GET `/api/workshop/`  
+   Lists all workshops that have surveys collecting responses.
+ * GET `/api/workshop/<workshop_id>/`  
+   Returns details of a particular workshop.
+ * GET `/api/workshop/<workshop_id>/survey/`  
+   Lists all the surveys whose responses are being collected.
+   Each survey has its corresponding `survey_id`.
+ * POST `/api/workshop/<workshop_id>/survey/refetch`  
+   Repopulates the list of surveys whose response are being
+   collected.
+ * GET `/api/workshop/<workshop_id>/survey/<survey_id>`  
+   Returns the responses for the particular survey, listed
+   according to the questions contained in the survey.
+ * POST `/api/workshop/<workshop_id>/survey/<survey_id>/refetch`  
+   Instructs the Django application to fetch the responses
+   that the survey has received since the last time the
+   SurveyMonkey API was hit.
+
+The underlying Django app will perform these tasks:
+ * Polling SurveyMonkey API management command:
+   This management command would populate the database with survey
+   responses fetched from the SurveyMonkey API.
+   The number of survey responses obtained from the [get_response_counts]
+   API call will be compared to the number of survey responses
+   saved in the database. The remaining responses will be fetched
+   using the [get_respondent_list] and [get_responses] API calls.
+ * Segregating the survey responses according to the answer to the
+   question: *Which workshop did you attend?* or the custom variable
+   `workshop_id`.
+ * Rendering and serving the templates.
+
+### Scheduler - Cron
+
+[Cron] will be used as a job scheduler to periodically run
+the *Polling SM API* management command at fixed time intervals.
+The interval duration would be dynamically determined by the
+number of API calls left and the time of the day.
+
+*Why Cron?*
+ * Cron is the easiest job scheduling solution available to us
+   and is available across all Unix environments.
+ * > Premature optimization is the root of all evil
+
+   -- DonaldKnuth
+
+### Database - PostgreSQL
+
+The Django app will use [PostgreSQL] as its database engine to
+store the survey responses fetched by the backend.
+
+*Why PostgreSQL?*
+ *  Django provides [PostgreSQL-specific features] like the ability
+    and query JSONs, dicts and arrays. The database schema, when
+    finalized, would benefit from these features.
+ *  PostgreSQL is actively maintained and is seen to perform
+    faster writes as compared to SQLite.
+ *  Although AMY currently uses SQLite database, a [migration]
+    is currently being planned to upgrade AMY to PostgreSQL.
```

Comment:
> We want to have:
 >  surveys
 >  surveys' questions
 >  responses grouped by workshop_id and by respondent_id

The database schema depend on how are we filtering responses by workshop:
* **Only the custom variable `workshop_id`:**
We do not want to shift to this as of now. However, this allows us to replace the *Workshop* and the *Survey_Workshop* tables with just one character column, `survey_id` on the *Survey* table.
* **Only the answer to the question *Please select the workshop you attended?":**
 An answer to a question is returned from SM API as:
```json
{
    "answers": [
        {
            "row": "3024959616"
        }
    ],
    "question_id": "316083321"
}
```
We can identify the question from the id `316083321`.
From a list(also called rows) of answers, the answer that was selected can be
identified with the id `3024959616`.
We need a way to associate the id `3024959616` with a particular workshop.
So we use a ForeignKey to a *Workshop* table.
But the pre-workshop survey may use a different id (not necessarily `3024959616`)
than the post-workshop survey to refer to the same workshop, as these workshops are added to the answer list using the SM web dashboard.
This is the reason why we need a through table with a column `answer_id` because
the relationship between a workshop and a survey is not simple.
* **Both, with higher priority given to one of them:**
This is similar to the above situation.